nemo/collections/nlp/models/machine_translation/transformer_mt_model.py

```diff
@@ -0,0 +1,332 @@
+
+import numpy as np
+import torch
+import torch.utils.data as pt_data
+from omegaconf import DictConfig
+from pytorch_lightning import Trainer
+from pytorch_lightning.utilities import rank_zero_only
+from sacrebleu import corpus_bleu
+
+from nemo.collections.common.losses import SmoothedCrossEntropyLoss
+from nemo.collections.common.parts import transformer_weights_init
+from nemo.collections.nlp.data import TranslationDataset
+from nemo.collections.nlp.modules.common import TokenClassifier
+from nemo.collections.nlp.modules.common.tokenizer_utils import get_tokenizer
+from nemo.collections.nlp.modules.common.transformer import (
+    BeamSearchSequenceGenerator,
+    TransformerDecoder,
+    TransformerEmbedding,
+    TransformerEncoder,
+)
+from nemo.core.classes.common import typecheck
+from nemo.core.classes.modelPT import ModelPT
+from nemo.utils import logging
+
+__all__ = ['TransformerMTModel']
+
+
+class TransformerMTModel(ModelPT):
+    """
+    Left-to-right Transformer language model.
+    """
+
+    def __init__(self, cfg: DictConfig, trainer: Trainer = None):
+        # shared params for dataset and data loaders
+        self.num_examples = {
+            "test": 3,
+            "val": 3,
+        }
+
+        if "tokenizer" in cfg.machine_translation:
+            if "src_tokenizer" in cfg.machine_translation or "tgt_tokenizer" in cfg.machine_translation:
+                raise ValueError(
+                    "If 'tokenizer' is in 'machine_translation' section of config then this section should "
+                    "not contain 'src_tokenizer' and 'tgt_tokenizer' fields."
+                )
+            self.src_tokenizer = get_tokenizer(**cfg.machine_translation.tokenizer)
+            self.tgt_tokenizer = self.src_tokenizer
+            super().__init__(cfg=cfg, trainer=trainer)
+            # make vocabulary size divisible by 8 for fast fp16 training
+            src_vocab_size = 8 * math.ceil(self.src_tokenizer.vocab_size / 8)
+            tgt_vocab_size = src_vocab_size
+            self.src_embedding_layer = TransformerEmbedding(
+                vocab_size=src_vocab_size,
+                hidden_size=cfg.machine_translation.hidden_size,
+                max_sequence_length=cfg.machine_translation.max_seq_length,
+                embedding_dropout=cfg.machine_translation.get("embedding_dropout", 0.0),
+                learn_positional_encodings=False,
+            )
+            self.tgt_embedding_layer = self.src_embedding_layer
+        else:
+            self.src_tokenizer = get_tokenizer(**cfg.machine_translation.src_tokenizer)
+            self.tgt_tokenizer = get_tokenizer(**cfg.machine_translation.tgt_tokenizer)
+            super().__init__(cfg=cfg, trainer=trainer)
+            # make vocabulary size divisible by 8 for fast fp16 training
+            src_vocab_size = 8 * math.ceil(self.src_tokenizer.vocab_size / 8)
+            tgt_vocab_size = 8 * math.ceil(self.tgt_tokenizer.vocab_size / 8)
```

Comment:
I'm not familiar with this idea for fast fp16 training, but with some very low probability, the model might output an invalid index that does not correspond to any vocab item.

Reply:
Great point! I added the filtering of beam search results. All ids greater or equal to vocabulary size are replaced with UNK id.